objiotracing/lib/simulate.go

```diff
@@ -0,0 +1,143 @@
+package lib
+
+import (
+	"fmt"
+
+	"github.com/cockroachdb/pebble/objstorage/objstorageprovider/objiotracing"
+	"github.com/dgryski/go-clockpro"
+	"github.com/dgryski/go-s4lru"
+	"github.com/dgryski/go-tinylfu"
+)
+
+type ReplacementPolicy int
+
+const (
+	ClockPro ReplacementPolicy = iota
+	S4LRU
+	TinyLFU
+	// TODO(josh): Implement.
+	LRU
+)
+
+type Config struct {
+	Policy                   ReplacementPolicy
+	// If nil, then (i) assume reads & writes will always be in units of pebble sstable
+	// block size and (ii) do caching in units of pebble sstable blocks.
+	// TODO(josh): I am unsure if the first assumption is okay to make. A related question
+	// I have is whether the existing in-memory pebble block cache caches in units of pebble
+	// sstable blocks.
+	BlockSize                *int64
+	CacheSize                int
+	// Must be set if Policy == TinyLFU. Else must be nil.
+	TinyLFUSamples           *int
+	WriteThru                bool
+	CacheUserFacingReadsOnly bool
+	L5AndL6Only              bool
+}
+
+type Results struct {
+	Hits   int
+	Misses int
+}
+
+type Cache interface {
+	Get(key string) interface{}
+	Set(key string, value interface{})
+}
+
+type wrappedS4LRU struct {
+	c *s4lru.Cache
+}
+
+func (c *wrappedS4LRU) Get(key string) interface{} {
+	val, ok := c.c.Get(key)
+	if !ok {
+		return nil
+	}
+	return val
+}
+
+func (c *wrappedS4LRU) Set(key string, value interface{}) {
+	c.c.Set(key, value)
+}
+
+type wrappedTinyLFU struct {
+	c *tinylfu.T
+}
+
+func (c *wrappedTinyLFU) Get(key string) interface{} {
+	val, ok := c.c.Get(key)
+	if !ok {
+		return nil
+	}
+	return val
+}
+
+func (c *wrappedTinyLFU) Set(key string, value interface{}) {
+	c.c.Add(key, value)
+}
+
+func Simulate(trace []objiotracing.Event, config Config) Results {
+	var results Results
+	var cache Cache
+	if config.Policy == ClockPro {
+		cache = clockpro.New(config.CacheSize)
+	} else if config.Policy == S4LRU {
+		cache = &wrappedS4LRU{s4lru.New(config.CacheSize)}
+	} else if config.Policy == TinyLFU {
+		if config.TinyLFUSamples == nil {
```

Comment:
I found a "driver" program from the same author and he sets the sample size to `10*CacheSize`. https://github.com/dgryski/trifles/blob/master/cachetest/main.go

That is consistent with Figure 8 in the [paper](https://arxiv.org/pdf/1512.00727.pdf), the first graph is on that order and the second graph mentions `20*CacheSize`.

Reply:
Thanks! In my next PR, I am setting sample size to 10x cache size.